examples/iot/esp32/voice_led_controller/led_controller.ino.ino

```diff
@@ -0,0 +1,433 @@
+/*
+Copyright 2025 Google LLC.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+#include <Arduino.h>
+#include <WiFi.h>
+#include <FS.h>
+#include <SD.h>
+#include <HTTPClient.h>
+#include <WiFiClientSecure.h>
+#include <ArduinoJson.h>
+#include "soc/soc.h"
+#include "soc/rtc_cntl_reg.h"
+#include <Adafruit_NeoPixel.h>
+
+// Pins
+const int SD_CS = 5;
+const int AUDIO_PIN = 34;
+const int BUTTON_PIN = 32;
+const int LED_PIN = 33;
+const int NEOPIXEL_PIN = 15
+
+// Configs for LED ring
+const int NEOPIXEL_COUNT = 24;
+int red = 255;
+int green = 255;
+int blue = 255;
+
+// Configuration for audio recording
+const int SAMPLE_RATE = 8000;
+const int BIT_DEPTH = 16;
+const int RECORD_DURATION = 2;
+
+Adafruit_NeoPixel pixels(NEOPIXEL_COUNT, NEOPIXEL_PIN, NEO_GRB + NEO_KHZ800);
+
+// WIFI connection
+String SSID = "";
+String PASSWORD = "";
+
+// Gemini API key
+String API_KEY = "";
+
+void setupWifi() {
+  WiFi.begin(SSID, PASSWORD);
+  while (WiFi.status()!= WL_CONNECTED) {
+    delay(1000);
+    Serial.print("...");
+  }
+  Serial.print("IP address: ");
+  Serial.println(WiFi.localIP());
+}
+
+void toggleLights(bool on) {
+  if( on ) {
+    Serial.println("Turning on lights");
+    int color = 255;
+    for (int i = 0; i < NEOPIXEL_COUNT; i++) {
+     pixels.setPixelColor(i, pixels.Color(red, green, blue));
+     pixels.setBrightness(255);
+    }
+    pixels.show();
+  } else {
+    Serial.println("Turning off lights");
+    pixels.clear();
+    pixels.show(); 
+  }
+}
+
+void recordAudio() {
+  if (!SD.begin(SD_CS, SPI, 1000000)) {
+    Serial.println("SD card initialization failed!");
+    while (1);
+  } else {
+    Serial.println("SD card initialized!");
+  }
+
+  if (SD.exists("/tmp.wav")) {
+    if (SD.remove("/tmp.wav")) {
+      Serial.println("Previous audio file deleted.");
+    } else {
+      Serial.println("Failed to delete previous audio file.");
+      return;
+    }
+  } else {
+    Serial.println("No previous audio file detected, starting new");
+  }
+
+  File audioFile = SD.open("/tmp.wav", FILE_WRITE);
+  if (!audioFile) {
+    Serial.println("Failed to create audio file.");
+    return;
+  }
+
+  Serial.println("Start recording");
+  writeWavHeader(audioFile, SAMPLE_RATE, BIT_DEPTH, 1);
+
+  int numSamples = SAMPLE_RATE * RECORD_DURATION;
+  for (int i = 0; i < numSamples; i++) {
```

Comment:
(no need to do anything now)

Does the streaming API work in this context at all? I'm curious if streaming small chunks of audio would work well within these memory constraints, and whether letting the model's VAD determine whether to respond would avoid the need for fixed-length and explicit audio turns.

Reply:
I'll have to look into it. That'd be a solid second demo though.